Rework BOM handling and encoding API (#399)
tafia/quick-xml: Changelog.md

- [#393]: Now `BytesStart::name()` and `BytesEnd::name()` returns `QName`, and
  `BytesStart::local_name()` and `BytesEnd::local_name()` returns `LocalName`

- [#191]: Remove unused `reader.decoder().decode_owned()`. If you ever used it,
  use `String::from_utf8` instead (which that function did)
- [#191]: Remove `*_without_bom` methods from the `Attributes` struct because they are useless.
  Use the same-named methods without that suffix instead. Attribute values cannot contain BOM

> - [#191]: Remove `Reader::decode()` and `Reader::decode_without_bom()`, they was replaced by

Reviewer: ```suggestion
- [#191]: Remove `Reader::decode()` and `Reader::decode_without_bom()`, they are replaced by
```

  `Decoder::decode()` and `Decoder::decode_with_bom_removal()`.
  Use `reader.decoder().decode_*(...)` instead of `reader.decode_*(...)` for now.
  `Reader::encoding()` was replaced by `Decoder::encoding()` as well
- [#191]: Remove poorly designed `BytesText::unescape_and_decode_without_bom()` and
  `BytesText::unescape_and_decode_without_bom_with_custom_entities()`. Although these methods worked
  as expected, this was only due to good luck. They was replaced by the
  `BytesStartText::decode_with_bom_removal()`:
  - conceptually, you should decode BOM only for the first `Text` event from the
    reader (since now `StartText` event is emitted instead for this)
  - text before the first tag is not an XML content at all, so it is meaningless
    to try to unescape something in it

> - [#180]: Define the same API for encoding whether feature `encoding` is defined or not.

Reviewer: Could this be replaced with "Since the encoding feature no longer performs automatic replacement for invalid characters."?

Author: I'm trying to say, that I've fixed the situation with different `decode*()` signatures when `encoding` feature enabled and when it is disable (returning `Cow<&str>` vs `Result<&str>`). After the fix both realizations returns `Result<Cow<&str>>`

  Since then no more replacements does for invalid characters when `encoding` feature

>   is enabled. An error will be returned instead, as for the case when `encoding` is disabled.

Reviewer: Here you say an error will be returned when `encoding` is disabled, but then below you say the opposite?

Author: You mean that:
> If `encoding` feature is disabled, `Cow::Borrowed` is returned instead of `&str`.
?
Yes, probably I should write the types completely. It now returns `Result<Cow<&str>>` (instead of `Result<&str>`), always with `Cow::Borrowed` variant

- [#191]: Remove `*_without_bom` methods from the `Attributes` struct because they are useless.
  Use the same-named methods without that suffix instead. Attribute values cannot contain BOM
- [#191]: Remove `Reader::decode()` and `Reader::decode_without_bom()`, they are replaced by
  `Decoder::decode()` and `Decoder::decode_with_bom_removal()`.
  Use `reader.decoder().decode_*(...)` instead of `reader.decode_*(...)` for now.
  `Reader::encoding()` is replaced by `Decoder::encoding()` as well
- [#191]: Remove poorly designed `BytesText::unescape_and_decode_without_bom()` and
  `BytesText::unescape_and_decode_without_bom_with_custom_entities()`. Although these methods worked
  as expected, this was only due to good luck. They was replaced by the
  `BytesStartText::decode_with_bom_removal()`:
  - conceptually, you should decode BOM only for the first `Text` event from the
    reader (since now `StartText` event is emitted instead for this)
  - text before the first tag is not an XML content at all, so it is meaningless
    to try to unescape something in it

- [#180]: Eliminate differences in the decoding API when feature `encoding` enabled and when it is

>   disabled. Since then signatures of functions are the same regardless of the feature enabled and

Reviewer: ```suggestions
  disabled. Signatures of functions are now the same regardless of whether the feature is enabled and
```

>   disabled. Since then signatures of functions are the same regardless of the feature enabled and

Reviewer: Eliminated the differences in the decoding API when feature `encoding` enabled and when it is
disabled. Signatures of functions are now the same regardless of whether or not the feature is
enabled, and an error will be returned instead of performing replacements for invalid characters
in both cases.

>   no more replacements does for invalid characters when feature is enabled. An error will be returned

Reviewer: ```suggestion
instead of performing replacements for invalid characters when the feature is enabled, as it is when `encoding` is disabled.
```

  instead, as for the case when `encoding` is disabled.

>   Namely, if `encoding` feature is disabled, `Result<Cow<&str>>` is returned instead of `Result<&str>`.

Reviewer: Previously, if the `encoding` feature was enabled, decoding functions would return `Result<Cow<&str>>` while without this feature they would return `Result<&str>`.  With this change, only `Result<Cow<&str>>` is returned regardless of the status of the feature.
